Extend Canvas TextMetrics for editing and text styling

AndresRPerez12 opened the pull request:
<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.

When you submit this PR, and each time you edit this comment (including checking a checkbox through the UI!), PR Preview will run and update it. As such make any edits in one go and only after PR Preview has run.

If you think your PR is ready to land, please double-check that the build is passing and the checklist is complete before pinging.
-->

This change adds the following functionalities to TextMetrics objects:
- Get the set of rectangles that the user agent would render as the selection background when a particular character range of the measured text is selected.
- Get the rectangle corresponding to the actual bounding box for a character range of the measured text. This is the equivalent to TextMetrics.actualBoundingBox restricted to the given range.
- Get the string index for the character at a given offset distance from the start position of the measured text.
- Split a character range of the measured text into grapheme clusters. That is, for the given range, it returns the minimal logical units of text, each of which can be rendered, along with their corresponding positional data. 

Additionally, a new method is added to the Canvas 2D rendering context to render these clusters. When rendering, all relevant CanvasTextDrawingStyles are used as they were when the text was measured.

See the discussion in the issue: https://github.com/whatwg/html/issues/10677

- [ ] At least two implementers are interested (and none opposed):
   * Chromium:  The new methods have been implemented (https://issues.chromium.org/u/1/issues/341213359)
   * Gecko: Position has been requested (https://github.com/mozilla/standards-positions/issues/1144)
   * WebKit: Position has been requested, and some portions have already started landing (https://github.com/WebKit/standards-positions/issues/436)
- [X] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://github.com/web-platform-tests/wpt/tree/master/html/canvas/element/text
   * https://github.com/web-platform-tests/wpt/tree/master/html/canvas/offscreen/text
   * Relevant tests: `2d.text.measure.selection-rects*`, `2d.text.measure.getActualBoundingBox*`, `2d.text.measure.index-from-offset*`, and `2d.text.measure.text-clusters*`.
- [ ] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: https://crbug.com/341213359
   * Gecko: …
   * WebKit: https://bugs.webkit.org/show_bug.cgi?id=284464
- [ ] Corresponding [HTML AAM](https://w3c.github.io/html-aam/) & [ARIA in HTML](https://w3c.github.io/html-aria/) issues & PRs:
- [ ] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: …
- [X] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. <!-- If you created this PR from a single commit, Github copied its message. Otherwise, you need to add a commit message yourself. -->

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)

schenney-chromium commented on source:
```diff
@@ -69143,6 +69175,12 @@ try {
+   <dt><code data-x=""><var>metrics</var>.<span subdfn data-x="dom-textmetrics-getIndexFromOffset">getIndexFromOffset</span>(<var>offset</var>)</code></dt>
+   <dt><code data-x=""><var>metrics</var>.<span subdfn data-x="dom-textmetrics-getTextClusters">getTextClusters</span>(<var>start</var>, <var>end</var> [, <var>options</var> ])</code></dt>
+
+   <dt><code data-x=""><var>context</var>.<span subdfn data-x="dom-context-2d-fillTextCluster">fillTextCluster</span>(<var>textCluster</var>, <var>x</var>, <var>y</var> [, <var>options</var> ])</code></dt>
```
The fill method belongs up with the other drawing methods.

AndresRPerez12 replied:
Done, thanks!

schenney-chromium commented on source:
```diff
@@ -65506,6 +65515,7 @@ interface mixin <dfn interface>CanvasText</dfn> {
   undefined <span data-x="dom-context-2d-fillText">fillText</span>(DOMString text, unrestricted double x, unrestricted double y, optional unrestricted double maxWidth);
   undefined <span data-x="dom-context-2d-strokeText">strokeText</span>(DOMString text, unrestricted double x, unrestricted double y, optional unrestricted double maxWidth);
   <span>TextMetrics</span> <span data-x="dom-context-2d-measureText">measureText</span>(DOMString text);
+  undefined <span data-x="dom-context-2d-fillTextCluster">fillTextCluster</span>(<span>TextCluster</span> textCluster, double x, double y, optional <span>TextClusterOptions</span> options);
```
I would expect this to be up with drawing methods.

AndresRPerez12 replied:
Done, thanks!

schenney-chromium commented on source:
```diff
@@ -65604,6 +65614,28 @@ interface <dfn interface>TextMetrics</dfn> {
+    readonly attribute <span>CanvasTextBaseline</span> <span data-x="dom-TextCluster-baseline">baseline</span>;
+};
+
+dictionary <dfn dictionary>TextClusterOptions</dfn> {
```
Looks like dictionaries go before their first use.

AndresRPerez12 commented:
Tagging: @whatwg/canvas

schenney-chromium commented on source:
```diff
@@ -69327,6 +69365,116 @@ try {
+        data-x="dom-context-2d-measureText">measureText()</code> was called) in relation to the text
+        as a whole.</p>
+
+        <p>The x position specified for each cluster corresponds to the aligment point given by the
```
Typo aligment

AndresRPerez12 replied:
Done, thanks!

schenney-chromium commented on source:
```diff
@@ -69327,6 +69365,116 @@ try {
+
+      <dt><dfn attribute for="TextCluster"><code data-x="dom-TextCluster-end">end</code></dfn> attribute</dt>
+
+      <dd><p>The ending index for the range of <span data-x="code point">code points</span> that are
```
Is this the index one past the end? Range APIs seem to behave that way but I'm not sure how conventional it is.

AndresRPerez12 replied:
Yes, it is one past the end. I hadn't realized I didn't clarify it in the spec. I'll add it explicitly as I don't know how conventional it is either.

schenney-chromium commented on source:
```diff
@@ -69327,6 +69365,116 @@ try {
+      for each cluster.</p></dd>
+    </dl>
+
+    <p class="note">The user agent might internally store in the <code>TextCluster</code> object the
+    complete text, as well as all the <code>CanvasTextDrawingStyles</code> at the time that <code
+    data-x="dom-context-2d-measureText">measureText()</code> was called, as they will be needed to
+    correctly render the clusters as they were measured.</p>
```
Is it a requirement that the drawing command uses the same parameters as the measurement? If so I would say that directly here, because right now it seems suggestive but not prescribed.

AndresRPerez12 replied:
Yes, that is specified in the section for `fillTextCluster()`. The idea when I first wrote it was to just mention the behavior expected, even if the implementation would most likely be to store the parameters in the cluster (as is implemented right now in Chrome).

But, following a suggestion by @Kaiido in another comment,  I changed the phrasing of that requirement by using `TextCluster` as an opaque object that stores those parameters in cluster when they are created. Hopefully this is more clear. Let me know what you think!

schenney-chromium reviewed (commented):
Sorry for leaving some comments out of the review.

Kaiido commented on source:
```diff
@@ -69339,7 +69487,67 @@ try {
+    pixels</span>.</p></li>
+
+    <li>
+      <p>If a <code>TextClusterOptions</code> options dictionary is passed and it has an <code
+      data-x="">options.x</code> value, move all the shapes in <var>glyphs</var> to the right by
+      <code data-x="">options.x-cluster.x</code>.</p>
+
+      <p>If a <code>TextClusterOptions</code> options dictionary is passed and it has an <code
+      data-x="">options.y</code> value, move all the shapes in <var>glyphs</var> down by <code
+      data-x="">options.y-cluster.y</code>.</p>
```
I think at this point `options` and `cluster` are [infra maps](https://infra.spec.whatwg.org/#ordered-map), and thus this should be `<var>options["x"]</var>` instead of `options.x`. Also I think the `-` should be `&minus;` in free text. So e.g. for the last one, that'd give `<var>options["y"]</var> &minus; <var>cluster["y"]</var>`, but here again take this with a grain of salt.

AndresRPerez12 replied:
I agree, thanks for pointing me in the direction of infra maps. I've updated here and elsewhere to reflect their use.

Kaiido reviewed (commented):
Thanks for doing this!

A couple of questions that came to mind:  

- Should there be a `strokeTextCluster()` method akin to `strokeText()`?
- Should the `TextCluster` be serializable? (so that we can send it to/from a Worker).

Also, I didn't receive the @whatwg/canvas notification from your previous comment. Hopefully this one will work better.

AndresRPerez12 commented on source:
```diff
@@ -69339,7 +69487,67 @@ try {
+  data-x="dom-context-2d-fillTextCluster">fillTextCluster(<var>textCluster</var>, <var>x</var>,
+  <var>y</var>, <var>options</var>)</code></dfn> method renders a <code>TextCluster</code> object.
+  The cluster is rendered where it would be if the whole text that was passed to <code
+  data-x="dom-context-2d-measureText">measureText()</code> to obtain the cluster was rendered at
```
I hadn't noticed that objects like `CanvasPattern` were defined in this way as opaque. I like that concept and I feel it applies well to this use case. Thank you so much for the suggestion!

I have updated it to reflect this. For now I went with the second approach you mentioned (storing the text and the `CanvasTextDrawingStyles` directly in the cluster as internal values). Hopefully this provides more clarity. Let me know what you think!

AndresRPerez12 commented on source:
```diff
@@ -69339,7 +69487,67 @@ try {
+  The cluster is rendered where it would be if the whole text that was passed to <code
+  data-x="dom-context-2d-measureText">measureText()</code> to obtain the cluster was rendered at
+  position (<var>x</var>,<var>y</var>), unless the positioning is modified with the options
+  argument. Specifically, when the methods are invoked, the user agent must run these steps:</p>
```
Thanks for catching this. That came from me using the algorithm for `fillText()` and `strokeText()` directly and not noticing this plural. Since I have added `strokeText()` in this update, now the plural form is the correct one.

AndresRPerez12 commented on source:
```diff
@@ -69339,7 +69487,67 @@ try {
+    <li><p>Run the <span>text preparation algorithm</span>, passing it the <var>text</var>
+    originally passed to <code data-x="dom-context-2d-measureText">measureText()</code> to obtain
+    the cluster, and an object with the <code>CanvasTextDrawingStyles</code> values as they were
+    when the cluster was measured, with the exception of <code
```
With the use of the internal values of the opaque object, hopefully the relationship is clearer now. Passing the exact object from the cluster won't always be possible due to the possibility of passing `align` and `baseline` to to `fillTextCluster()` when rendering, but most of the values now come from the internally stored value.

AndresRPerez12 commented on source:
```diff
@@ -69339,7 +69487,67 @@ try {
+    data-x="dom-TextCluster-baseline">baseline</code> attributes of <var>cluster</var>. Let
+    <var>glyphs</var> be the result.</p></li>
+
+    <li><p>Filter <var>glyphs</var> to include only glyphs that contain <span
```
Done, thanks!

AndresRPerez12 commented:
Thank you both for your comments! I have addressed them and uploaded the changes. Please let me know what you think.

Answering some of the open questions:

> * Should there be a `strokeTextCluster()` method akin to `strokeText()`?

I have added it to the PR.  I agree that it makes sense to have both.

> * Should the `TextCluster` be serializable? (so that we can send it to/from a Worker).

We feel that for now, it's simpler to not make it serializable. Once we merge this and we know exactly what clusters look like and what values they are storing internally that would require serialization, we can look into it.

> Also, I didn't receive the @whatwg/canvas notification from your previous comment. Hopefully this one will work better.

 Thanks! I'm not fully sure how the tags work, so I really appreciate it!

annevk commented on source:
```diff
@@ -4165,6 +4165,7 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
          <dfn data-x="DOMPointInit-x" data-xref="https://drafts.fxtf.org/geometry/#dom-dompointinit-x">x</dfn> and
          <dfn data-x="DOMPointInit-y" data-xref="https://drafts.fxtf.org/geometry/#dom-dompointinit-y">y</dfn> members</li>
       <li><dfn data-x-href="https://drafts.fxtf.org/geometry/#matrix-multiply">Matrix multiplication</dfn></li>
+      <li><dfn data-x-href="https://www.w3.org/TR/geometry-1/#DOMRect">DOMRectReadOnly and DOMRect</dfn></li>
```
Each needs its own reference.

AndresRPerez12 replied:
I added both since the section on W3C is named `DOMRect`, but since I’m only using `DOMRectReadOnly` I will just leave that one pointing to its definition.

domenic replied:
Please do not use TR links; use https://drafts.fxtf.org/geometry/ .

annevk replied:
And objects should have `<code>` as well.

annevk commented on source:
```diff
@@ -69196,6 +69254,77 @@ try {
    </li>
   </ol>
 
+  <p>The <dfn method for="CanvasText"><code
+  data-x="dom-context-2d-fillTextCluster">fillTextCluster(<var>cluster</var>, <var>x</var>,
+  <var>y</var>, <var>options</var>)</code></dfn> and <dfn method for="CanvasText"><code
+  data-x="dom-context-2d-strokeTextCluster">strokeTextCluster(<var>cluster</var>, <var>x</var>,
+  <var>y</var>, <var>options</var>)</code></dfn> methods render a <code>TextCluster</code> object.
+  The cluster is rendered where it would be if the whole text that was passed to <code
+  data-x="dom-context-2d-measureText">measureText()</code> to obtain the cluster was rendered at
+  position (<var>x</var>,<var>y</var>), unless the positioning is modified with the options
+  argument. Specifically, when the method is invoked, the user agent must run these steps:</p>
```
These need to be rephrased to use the modern way of defining methods. "The X method steps are ..."

AndresRPerez12 replied:
Done, although I kept the structure of bundling the fill and stroke version together.

If this version is what is desired, I can create a PR moving `fillText()` and `strokeText()` to this modern format once this PR lands.

annevk commented on source:
```diff
@@ -69196,6 +69254,77 @@ try {
+  position (<var>x</var>,<var>y</var>), unless the positioning is modified with the options
+  argument. Specifically, when the method is invoked, the user agent must run these steps:</p>
+
+  <ol>
+    <li><p>If any of the arguments are infinite or NaN, then return.</p></li>
```
Indentation and formatting here doesn't appear to follow the guidelines.

annevk commented on source:
```diff
@@ -69327,6 +69456,125 @@ try {
    baseline</span>. (Zero if the given baseline is the <span>ideographic-under
    baseline</span>.)</p></dd>
+
+   <dt><dfn method for="TextMetrics"><code data-x="dom-textmetrics-getSelectionRects">getSelectionRects(<var>start</var>, <var>end</var>)</code></dfn> method</dt>
```
This should not follow the bad precedent set of the current set of attributes. Please define these as modern methods.

annevk replied:
I don't understand how this relates to the concern?

AndresRPerez12 replied:
Sorry my reply wasn’t clear. I did modernize these methods, but since their behavior depends heavily on the result of the text preparation algorithm when `measureText()` was called before, hence my previous comment wondering if you agree with the way I referenced those returned values in order to modernize the phrasing.

In this round of updates I have also modernized `getIndexFromOffset()`.

annevk commented on source:
```diff
@@ -65604,6 +65615,22 @@ interface <dfn interface>TextMetrics</dfn> {
+    // opaque object
+    readonly attribute double <span data-x="dom-TextCluster-x">x</span>;
+    readonly attribute double <span data-x="dom-TextCluster-y">y</span>;
+    readonly attribute unsigned long <span data-x="dom-TextCluster-begin">begin</span>;
```
Sometimes you use "begin" and sometimes "start". That should be made consistent.

AndresRPerez12 replied:
Done, I settled with start. Thanks for pointing it out!

annevk commented on source:
```diff
@@ -65604,6 +65615,22 @@ interface <dfn interface>TextMetrics</dfn> {
+
+[Exposed=(Window,Worker)]
+interface <dfn interface>TextCluster</dfn> {
+    // opaque object
```
I'm not sure what this comment buys us.

AndresRPerez12 replied:
I wanted to show that `TextCluster` objects have hidden valued stored inside that are defined when creating the cluster, and used when rendering it.  I added that comment since I saw it on `CanvasGradient` and `CanvasPattern`, that have this kind of hidden attributes. Should I express this in a different way?

annevk replied:
I see, if it's consistent with `CanvasPattern` we can keep it I suppose.

annevk commented on source:
```diff
@@ -65501,10 +65502,20 @@ interface mixin <dfn interface>CanvasUserInterface</dfn> {
+  <span>CanvasTextAlign</span> align;
+  <span>CanvasTextBaseline</span> baseline;
+  double x;
+  double y;
```
These members should probably be `<dfn>` so they can be referenced from prose.

AndresRPerez12 replied:
Done! I also referenced these new definitions where possible.

annevk commented on source:
```diff
@@ -69196,6 +69254,73 @@ try {
+  data-x="dom-context-2d-fillTextCluster">fillTextCluster(<var>cluster</var>, <var>x</var>,
+  <var>y</var>, <var>options</var>)</code></dfn> and <dfn method for="CanvasText"><code
+  data-x="dom-context-2d-strokeTextCluster">strokeTextCluster(<var>cluster</var>, <var>x</var>,
+  <var>y</var>, <var>options</var>)</code></dfn> methods, when invoked, must run these steps:</p>
```
"The X method steps are:" is the modern style.

AndresRPerez12 replied:
Done, thank you.

annevk commented on source:
```diff
@@ -69196,6 +69254,73 @@ try {
+  <var>y</var>, <var>options</var>)</code></dfn> methods, when invoked, must run these steps:</p>
+
+  <ol>
+   <li><p>If any of the arguments are infinite or NaN, then return.</p></li>
```
Given that these also take a dictionary, I think we should be more explicit here.

AndresRPerez12 replied:
Agreed, done.

annevk commented on source:
```diff
@@ -69196,6 +69254,73 @@ try {
+
+   <li><p>Run the <span>text preparation algorithm</span>, passing it the complete text and the
+   <code>CanvasTextDrawingStyles</code> from the opaque <code>TextCluster</code>
+   <var>cluster</var>, with the exception of <code
```
This exception also refers to the cluster. So I don't really follow this.

This is also talking about public API whereas we shouldn't be touching that here.

annevk commented on source:
```diff
@@ -69196,6 +69254,73 @@ try {
+   <li><p>Filter <var>glyphs</var> to include only the glyphs that contain <span
+   data-x="code point">code points</span> within the range <code
+   data-x="dom-TextCluster-start">cluster["start"]</code> to <code
+   data-x="dom-TextCluster-end">cluster["end"]</code>.</p></li>
```
A glyph contains a code point?

AndresRPerez12 replied:
You’re right, it doesn’t make sense. I rephrased it here (and in getTextCluster() since I used the same phrasing). Let me know what you think.

annevk commented on source:
```diff
@@ -65501,10 +65502,20 @@ interface mixin <dfn interface>CanvasUserInterface</dfn> {
   // text (see also the <span>CanvasPathDrawingStyles</span> and <span>CanvasTextDrawingStyles</span> interfaces)
   undefined <span data-x="dom-context-2d-fillText">fillText</span>(DOMString text, unrestricted double x, unrestricted double y, optional unrestricted double maxWidth);
   undefined <span data-x="dom-context-2d-strokeText">strokeText</span>(DOMString text, unrestricted double x, unrestricted double y, optional unrestricted double maxWidth);
+  undefined <span data-x="dom-context-2d-fillTextCluster">fillTextCluster</span>(<span>TextCluster</span> cluster, double x, double y, optional <span>TextClusterOptions</span> options);
```
Optional dictionaries must default to `= {}`

AndresRPerez12 replied:
Done! Thanks for pointing it out.

annevk commented on source:
```diff
@@ -69196,6 +69254,73 @@ try {
+   <li>
+    <p>If a <code>TextClusterOptions</code> options dictionary is passed and it has an <code
+    data-x="">options["x"]</code> value, move all the shapes in <var>glyphs</var> to the right by
+    <code data-x="">options["x"] &minus; cluster["x"]</code>.</p>
```
The dictionary is always passed (see other comment). But this language and markup needs updating. I recommend looking at something recent, such as `TogglePopoverOptions`.

AndresRPerez12 replied:
Thanks for that reference. I updated the language and used the `exist` reference. Let me know what you think!

annevk commented on source:
```diff
@@ -69196,6 +69254,73 @@ try {
+     style</span> must be applied to the result of <span data-x="trace a path">tracing</span> the
+     shapes using the object implementing the <code>CanvasText</code> interface for the line
+     styles, and <span>this</span>'s <span data-x="concept-CanvasFillStrokeStyles-fill-style">fill
+     style</span> must be ignored.</p>
```
I think it would be cleaner if there was a shared algorithm that was passed either "fill" or "stroke" or some such as an internal enum value that you could branch on here.

AndresRPerez12 replied:
Done, I followed the example of the text preparation algorithm. Let me know if this is what you wanted. If so, once this is merged I’ll create another PR for doing the same with `fillText()` and `strokeText()`.

annevk commented on source:
```diff
@@ -69339,8 +69609,6 @@ try {
   dedicated way of doing multiline layout.</p>
 
-
-
```
Probably best not to remove those here.

annevk commented:
Adding "do not merge yet" as it seems this requires further issue discussion: https://github.com/whatwg/html/issues/10677#issuecomment-2787797845.
